librad/src/git/storage/read.rs

```diff
@@ -0,0 +1,347 @@
+// Copyright © 2019-2020 The Radicle Foundation <hello@radicle.foundation>
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::{convert::TryFrom, fmt::Debug, marker::PhantomData, path::Path};
+
+use git_ext::{self as ext, blob, is_not_found_err, RefLike, RefspecPattern};
+use std_ext::result::ResultExt as _;
+use thiserror::Error;
+
+use crate::{
+    git::types::{reference, Many, One, Reference},
+    identities::git::{Identities, Urn},
+    paths::Paths,
+    peer::PeerId,
+};
+
+use super::{
+    config::{self, Config},
+    glob::{self, Pattern},
+};
+
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum Error {
+    #[error(transparent)]
+    Config(#[from] config::Error),
+
+    #[error("malformed URN")]
+    Ref(#[from] reference::FromUrnError),
+
+    #[error(transparent)]
+    Blob(#[from] ext::blob::Error),
+
+    #[error(transparent)]
+    Git(#[from] git2::Error),
+}
+
+/// Low-level operations on the link "monorepo".
+pub struct Storage {
+    pub(super) backend: git2::Repository,
+    pub(super) peer_id: PeerId,
+}
+
+impl Storage {
+    /// Open the [`Storage`], initialising it if it doesn't exist.
+    ///
+    /// Note that a [`Storage`] is tied to the [`Signer`] with which it was
+    /// initialised, attempting to open it with a different one (that is, a
+    /// different key) will return an error.
+    ///
+    /// # Concurrency
+    ///
+    /// [`Storage`] can be sent between threads, but it can't be shared between
+    /// threads. _Some_ operations are safe to perform concurrently in much
+    /// the same way two `git` processes can access the same repository.
+    /// However, if you need multiple [`Storage`]s to be shared between
+    /// threads, use a [`Pool`] instead.
+    pub fn open(paths: &Paths) -> Result<Self, Error> {
+        crate::git::init();
+
+        let backend = match git2::Repository::open_bare(paths.git_dir()) {
+            Err(e) if is_not_found_err(&e) => {
```

Comment:
We can’t initialise the repo without a signer. Well, maybe we can with just a peer id? To use it rw, you need to bring the secret part.